Clinical-Genomics/scout
# Refactor LoqusDB settings parsing and introducing Loqus queries via REST API

This PR adds support to queries to loqus using REST APIs. Fix #2295

Note that the connection to Loqusdb at the moment is a really messy story and I think it needs fixing as a Scout major release fix, see a suggestion in #2357

In this PR (**that will not make it necessary to release a major**), I'm introducing a warning message if a user provides LoqusDB settings that I think should be **deprecated**, for instance:

```
LOQUSDB_SETTINGS = {
  "binary_path":"/Users/username/miniconda3/envs/loqus/bin/loqusdb",
  "config_path":"/Users/username/Documents/work/GITs/loqusdb/config.yaml"
}
```

or:

```
LOQUSDB_SETTINGS = [ 
  {
    "binary_path":"/Users/username/miniconda3/envs/loqus/bin/loqusdb",
    "config_path":"/Users/username/Documents/work/GITs/loqusdb/config.yaml"
  },
  {
    "id": "cancer_loqus",
    "binary_path":"/Users/username/miniconda3/envs/loqus/bin/loqusdb2",
    "config_path":"/Users/username/Documents/work/GITs/loqusdb/config2.yaml"
  }
]
```
**Note that these settings will still work. I just want to scare users into using this type of config:**

```
LOQUSDB_SETTINGS = {
  "default": {
       "binary_path":"/Users/username/miniconda3/envs/loqus/bin/loqusdb",
       "config_path":"/Users/username/Documents/work/GITs/loqusdb/config.yaml"
   },
   "custom_other_loqus" : {
       "binary_path":"/Users/username/miniconda3/envs/loqus/bin/loqusdb2",
       "config_path":"/Users/username/Documents/work/GITs/loqusdb/config2.yaml"
    },
    "custom_loqus_api": {
       "api_url" : ""http://127.0.0.1:9000"
    }
}
```

Note that the config above has 3 different loqus instances. **This PR introduces support for the latter type (REST APIs).**

**Prepare for testing**
1. clone and install in a dedicated environment [loqusDB executable](https://github.com/moonso/loqusdb), without setting up its demo database.
1. clone (but don't install) the [loqusdbapi](https://github.com/Clinical-Genomics/loqusdbapi)
1. ` cd loqusdbapi`
1. Restore the loqusdbapi database in a local mongo instance (localhost, port 27017) so that it can be used by loqusdb executable:
```mongorestore --host localhost:27017 -d loqusdb dump/```
This way both instances will use the same database, only the loqusapi will dump this database in a mongo container that will be started from docker-compose
1. Start the containers for the loqusdb api and its database by running `docker-compose up`
1. Make sure l**oqus executable works** by running the following commands (while in the right environment):
 - `loqusdb variants --variant-id 1_879537_T_C --case-count` (SNV)
 - `loqusdb variants -t sv -c 1 -s 7889972 -e 7890026 --end-chromosome 1 --sv-type DEL --case-count` (SV)
 - `loqusdb cases --count -t snv`
 7. Make sure **loqus container is up** and ready to receive the same queries:
- `curl -X GET "http://127.0.0.1:9000/variants/1_879537_T_C"` (SNV)
- `curl -X GET "http://127.0.0.1:9000/svs/?chrom=1&end_chrom=1&pos=7889972&end=7890026&sv_type=DEL"` (SV)
- `curl -X GET "http://127.0.0.1:9000/cases"`

8. Modify a scout SNV in the database to match one if the SNVs in loqus, for instance take [this variant](http://localhost:5000/cust000/643594/4836caf41faea5964bbbedfff5185257): 
  Just change the "position" value in the database to: 879537
9. Do the same with a structural variant. Keep in mind that the SV fields to match the loqus variant are chromosome, end_chrom, position, end, type of SV


**How to test**:
- [ ]  Test with one single instance of Loqusdb (the executable): modify the config file to contain this type of (to be) deprecated settings reflecting one instance in a binary file. For instance these are my settings:
 ```
LOQUSDB_SETTINGS = {
  "binary_path":"/Users/chiararasi/miniconda3/envs/loqus/bin/loqusdb",
  "config_path":"/Users/chiararasi/Documents/work/GITs/loqusdb/config.yaml"
}
```
- You should be able to see one loqus observation when you click on the SNV and the SV
- [ ] Test with one single instance of Loqusdb (APIs). Change config settings by adding this:
```
LOQUSDB_SETTINGS = {
  "default": {"api_url": "http://127.0.0.1:9000"}
}
```
- You should be able to see one observation when you click on the SNV and the SV
- [ ] Test with one instance of Loqus exec and one instance APIs. Change the loqus setting like this:
```
LOQUSDB_SETTINGS = {
  "default": {"api_url": "http://127.0.0.1:9000"},
  "loqus_exe": {
     "binary_path":"/Users/chiararasi/miniconda3/envs/loqus/bin/loqusdb",
     "config_path":"/Users/chiararasi/Documents/work/GITs/loqusdb/config.yaml"
   }
} 
```
- Check that in institute settings you can see and select the 2 loqus instances and you can switch between the 2:
![image](https://user-images.githubusercontent.com/28093618/107932149-84878080-6f7d-11eb-85b3-acb0341dc5d9.png)
- Check that the 2 variants contain the observations, regardless of which instance of loqus is selected

- [ ] Install this branch on Scout stage and make sure that variant observations work as expected.
 
**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
